Doc about migrating from v1beta1 to v1 (#5415)
tektoncd/pipeline: docs/migrating-v1beta1-to-v1.md

<!--
---
linkTitle: "Migrating from Tekton v1beta1"
weight: 4000
---
-->

# Migrating From Tekton `v1beta1` to Tekton `v1`

- [Changes to fields](#changes-to-fields)
- [Upgrading `PipelineRun.Timeout` to `PipelineRun.Timeouts`](#upgrading-pipelinerun.timeout-to-pipelinerun.timeouts)
- [Deprecating Resources from Task, TaskRun, Pipeline and PipelineRun](#deprecating-resources-from-task,-taskrun,-pipeline-and-pipelinerun)
- [Replacing `taskRef.bundle` and `pipelineRef.bundle` with Bundle Resolver](#replacing-taskRef.bundle-and-pipelineRef.bundle-with-bundle-resolver)

This document describes the differences between `v1beta1` Tekton entities and their
`v1` counterparts. It also describes the changed fields and the deprecated fields into v1.
## Changes to fields

In Tekton `v1`, the following fields have been changed:

| Old field | Replacement |
| --------- | ----------|

> | `PipelineRun.Timeout`| `timeouts.pipeline` |
> | `spec.taskRunSpecs.taskServiceAccountName` | `spec.taskRunSpecs.serviceAccountName` |
> | `spec.taskRunSpecs.taskPodTemplate` | `spec.taskRunSpecs.podTemplate` |
> | `status.taskResults` | `status.results` |
> | `status.pipelineResults` | `status.results` |

Reviewer: ```suggestion
| `pipelineRun.spec.timeout`| `pipelineRun.spec.timeouts.pipeline` |
| `pipelineRun.spec.taskRunSpecs.taskServiceAccountName` | `pipelineRun.spec.taskRunSpecs.serviceAccountName` |
| `pipelineRun.spec.taskRunSpecs.taskPodTemplate` | `pipelineRun.spec.taskRunSpecs.podTemplate` |
| `taskRun.status.taskResults` | `taskRun.status.results` |
| `pipelineRun.status.pipelineResults` | `pipelineRun.status.results` |
```

| `taskRun.spec.taskRef.bundle` | `taskRun.spec.taskRef.resolver` |
| `pipelineRun.spec.pipelineRef.bundle` | `pipelineRun.spec.pipelineRef.resolver` |
| `task.spec.resources` | removed from `Task` |
| `taskrun.spec.resources` | removed from `TaskRun` |
| `pipeline.spec.resources` | removed from `Pipeline` |
| `pipelineRun.spec.resources` | removed from `PipelineRun` |

## Deprecating `resources` from Task, TaskRun, Pipeline and PipelineRun

> Due to the deprecation of `PipelineResources`, the `resources` fields of Task, TaskRun, Pipeline and PipelineRun can be replaced by `Task` following the [reference](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineresources.md)

Reviewer: ```suggestion
`PipelineResources` are deprecated, and the `resources` fields of Task, TaskRun, Pipeline and PipelineRun has been removed. Please use `Tasks` instead. For more information, see [Replacing PipelineResources](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineresources.md)
```

## Replacing `taskRef.bundle` and `pipelineRef.bundle` with Bundle Resolver
Bundle resolver in remote resolution should be used instead of `taskRun.spec.taskRef.bundle` and `pipelineRun.spec.pipelineRef.bundle`

```yaml

> # TaskRun.yaml (v1beta1)

Reviewer: ```suggestion
apiVersion: tekton.dev/v1beta1
kind: TaskRun
```

spec:
  taskRef:
    name: example-task
    bundle: docker.io/ptasci67/example-oci@sha256:03

> # TaskRun.yaml (v1)

Reviewer: ```suggestion
---
apiVersion: tekton.dev/v1
kind: TaskRun
```

spec:
  taskRef:
    resolver: bundles
    params:
    - name: bundle

>       value: foo.bar/baz

Reviewer: can you please use a more realistic example here and above, and make sure they match?

    - name: name
      value: taskName
    - name: kind
      value: Task
```

## Replacing ClusterTask with Remote Resolution

> [Remote resolution](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resource-resolution.md) should be used instead if you're currently using ClusterTask CRD, which shall be deprecated.

Reviewer: ```suggestion
`ClusterTask` is deprecated. Please use the `cluster` resolver instead, for example: .... For more information, see [Remote resolution](https://github.com/tektoncd/community/blob/main/teps/0060-remote-resource-resolution.md).
```

| `taskRun.spec.taskRef.bundle` | `taskRun.spec.taskRef.resolver` |
| `pipelineRun.spec.pipelineRef.bundle` | `pipelineRun.spec.pipelineRef.resolver` |
| `task.spec.resources` | removed from `Task` |
| `taskrun.spec.resources` | removed from `TaskRun` |
| `pipeline.spec.resources` | removed from `Pipeline` |
| `pipelineRun.spec.resources` | removed from `PipelineRun` |

## Deprecating `resources` from Task, TaskRun, Pipeline and PipelineRun
`PipelineResources` are deprecated, and the `resources` fields of Task, TaskRun, Pipeline and PipelineRun has been removed. Please use `Tasks` instead. For more information, see [Replacing PipelineResources](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineresources.md)

## Replacing `taskRef.bundle` and `pipelineRef.bundle` with Bundle Resolver
Bundle resolver in remote resolution should be used instead of `taskRun.spec.taskRef.bundle` and `pipelineRun.spec.pipelineRef.bundle`.

The [`enable-bundles-resolver`](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#customizing-the-pipelines-controller-behavior) feature flag must be enabled to use this feature.

```yaml
apiVersion: tekton.dev/v1beta1

> kind: Taskrun

Reviewer: this comment and the one below were not resolved

> kind: Taskrun

Author: This might be we changed this to `TaskRun.yaml` and removed the `kind`.

spec:
  taskRef:
    name: example-task
    bundle: docker.io/ptasci67/example-oci@sha256:053a6cb9f3711d4527dd0d37ac610e8727ec0288a898d5dfbd79b25bcaa29828
---
apiVersion: tekton.dev/v1beta1
kind: Taskrun
spec:
  taskRef:
    resolver: bundles
    params:
    - name: bundle
      value: docker.io/ptasci67/example-oci@sha256:053a6cb9f3711d4527dd0d37ac610e8727ec0288a898d5dfbd79b25bcaa29828
    - name: name
      value: taskName
    - name: kind
      value: Task
```

## Replacing ClusterTask with Remote Resolution
`ClusterTask` is deprecated. Please use the `cluster` resolver instead, for example: .... 

> # TODO add example after #5404 is merged

Reviewer: I don't think this comment was resolved.

- [Upgrading `PipelineRun.Timeout` to `PipelineRun.Timeouts`](#upgrading-pipelinerun.timeout-to-pipelinerun.timeouts)
- [Replacing `Pipeline.Tasks[].Conditions` with `WhenExpressions`](#replacing-pipeline.tasks[].conditions-with-whenexpressions)
- [Deprecating Resources from Task, TaskRun, Pipeline and PipelineRun](#deprecating-resources-from-task,-taskrun,-pipeline-and-pipelinerun)
- [Replacing `taskRef.bundle` and `pipelineRef.bundle` with Bundle Resolver](#replacing-taskRef.bundle-and-pipelineRef.bundle-with-bundle-resolver)

This document describes the differences between `v1beta1` Tekton entities and their
`v1` counterparts. It also describes the changed fields and the deprecated fields into v1.
## Changes to fields

In Tekton `v1`, the following fields have been changed:

| Old field | New field |
| --------- | ----------|
| `spec.taskRunSpecs.taskServiceAccountName` | `spec.taskRunSpecs.serviceAccountName` |
| `spec.taskRunSpecs.taskPodTemplate` | `spec.taskRunSpecs.podTemplate` |
| `status.taskResults` | `status.results` |
| `status.pipelineResults` | `status.results` |
| `taskRun.spec.taskRef.bundle` | removed from `TaskRun` |

> | `pipelineRun.spec.pipelineRef.bundle` | removed from `PipelineRun` |

Reviewer: I know you mention below there's a replacement, but we should include that on this table so people don't think we've removed this functionality entirely. I would call the right column "replacement" and put in `taskRun.spec.taskRef.resolver` instead.

## Replacing `PipelineRunCancelled` status with `Cancelled`

> `Cancelled` status should be used for replacing the deprecated `PipelineRunCancelled`.

Reviewer: We've already removed `PipelineRunCancelled`.

> ## Upgrading `PipelineRun.Timeout` to `PipelineRun.Timeouts`

Reviewer: Can you put this in the chart above instead? Also, I would specifically mention that `timeouts.pipeline` is the replacement for `timeout`

If you are using `pipelineRun.spec.timeout` you must now use `pipelineRun.spec.timeouts` as `pipelineRun.spec.timeout` is deprecated. 

> ## Replacing `Pipeline.Tasks[].Conditions` with `WhenExpressions`

Reviewer: We've already removed support for conditions, so this block can be removed.

You must use `WhenExpressions` if you are using `Pipeline.Tasks[].Conditions`

```yaml
# Pipeline.yaml (v1)
  pipelineSpec:
    tasks:
      - name: create-file
        when:
          - input: "$(params.path)"
            operator: in
            values: ["README.md"]
```

> ## Deprecating `resources` from Task, TaskRun, Pipeline and PipelineRun

Reviewer: Can you put this in the chart too? I think we'll want to pull [this section](https://github.com/tektoncd/pipeline/blob/main/docs/migrating-v1alpha1-to-v1beta1.md#replacing-pipelineresources-with-tasks) into its own md file and link to it from both migration docs.

Author: Thanks Lee. I've pulled out the section in https://github.com/tektoncd/pipeline/pull/5423/. Please help take a look.

Reviewer: I think this actually makes more sense as part of the same PR.

The `resources` fields of Task, TaskRun, Pipeline and PipelineRun should not be used due to the deprecation of `PipelineResources`.

## Replacing `taskRef.bundle` and `pipelineRef.bundle` with Bundle Resolver

> Bundle resolver in remote resolution should be used instead of `taskRun.spec.taskRef.bundle` and `pipelineRun.spec.pipelineRef.bundle`

Reviewer: This also requires enabling the `enable-bundles-resolver` feature flag; please link to https://github.com/tektoncd/pipeline/blob/main/docs/install.md#customizing-the-pipelines-controller-behavior

@abayer would you mind adding the new resolvers flags to that section?

```yaml
# TaskRun.yaml (v1beta1)
spec:
  taskRef:
    name: example-task
    bundle: docker.io/ptasci67/example-oci@sha256:03

# TaskRun.yaml (v1)
spec:
  taskRef:
    resolver: bundles
    params:
    - name: bundle
      value: foo.bar/baz
    - name: name
      value: taskName
    - name: kind
      value: Task
```

> ## Replacing ClusterTask with Remote Resolution

Reviewer: Can you give an example of how a cluster task can be replaced with the cluster resolver?

Reviewer 2: Good call, PR incoming.

Reviewer 2: ha, wait, I thought I was responding to https://github.com/tektoncd/pipeline/pull/5415#discussion_r961027785 🤦 I'll wait for this and #5405 to land and then open a PR to do that. =)
